// markdown2html-less/lib/markdown2htmlLess.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = require("lodash");
const render_1 = require("./render/render");
require("./utils/json");
const log_1 = require("./utils/log");
class Markdown2HtmlLess {
    constructor(options = {}) {
        this.defaultOptions = {
            abbr: true,
            anchor: true,
            attrs: true,
            debug: false,
            deflist: true,
            emoji: true,
            expandTabs: true,
            footnote: true,
            highlightjs: true,
            ins: true,
            lazyHeaders: true,
            mark: true,
            mermaid: true,
            sub: true,
            sup: true,
            taskLists: true,
        };
        if (typeof options !== 'object') {
            throw Error('second argument must be an object');
        }
        options = options || {};
        lodash_1.defaults(options, this.defaultOptions);
        this.options = options;
        return this;
    }
    markdown2html(markdown) {
        if (typeof markdown !== 'string') {
            throw Error('first argument must be a string');
        }
        this.log(this.banner());
        const markdownRenderOptions = {
            abbr: this.options.abbr,
            anchor: this.options.anchor,
            attrs: this.options.attrs,
            debug: this.options.debug,
            deflist: this.options.deflist,
            emoji: this.options.emoji,
            expandTabs: this.options.expandTabs,
            footnote: this.options.footnote,
            highlightjs: this.options.highlightjs,
            ins: this.options.ins,
            lazyHeaders: this.options.lazyHeaders,
            mark: this.options.mark,
            mermaid: this.options.mermaid,
            sub: this.options.sub,
            sup: this.options.sup,
            taskLists: this.options.taskLists,
        };
        const render = new render_1.MarkdownRender(markdownRenderOptions);
        const html = render.renderToHtml(markdown);
        if (this.options.debug) {
            html.body = this.debugHeader() + '\n' + html.body;
        }
        return html;
    }
    log(msg) {
        if (this.options.debug) {
            log_1.Log.print(msg);
        }
    }
    banner() {
        let banner = '\n\n' + 'markdown2html-pro';
        banner += 'Parse markdown into HTML and add syntax highlighting';
        return banner;
    }
    debugHeader() {
        const info = require('../package.json') || {};
        const debugHeader = '<!--' +
            ' this HTML was generated using markdown2html-pro version ' +
            info.version +
            '.' +
            ' see an issue? file at ' +
            info.issuesUrl +
            '.' +
            ' please include the version in your issue. thanks for using markdown2html-pro!' +
            ' to learn more, visit ' +
            info.repositoryUrl +
            '.' +
            '  -->';
        return debugHeader;
    }
}
exports.Markdown2HtmlLess = Markdown2HtmlLess;
