// markdown2html-less/lib/render/render.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
class MarkdownRender {
    constructor(options) {
        this.options = options || {};
        this.modules = {};
        this.mermaidOptions = {};
        return this;
    }
    renderToHtml(mdContent) {
        this.loadModules();
        const renderer = this.getRenderer();
        const b = renderer.render(mdContent);
        const h = this.mermaidOptions.returnHead || '';
        return {
            body: b,
            head: h,
        };
    }
    getInitOptions(markdownContent) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.options.mermaid) {
                if (typeof markdownContent !== 'string') {
                    throw Error('first argument must be a string');
                }
                const defaultRootWebPath = path.join(__dirname, '..');
                const options = {
                    rootWebPath: defaultRootWebPath,
                };
                const cms = yield this.modules.mermaid.mermaid_pro_plugin_init_everytime(markdownContent);
                return cms;
            }
            return [];
        });
    }
    loadModules() {
        this.modules.MarkdownIt = require('markdown-it');
        if (this.options.taskLists) {
            this.modules.taskLists = require('markdown-it-task-lists');
        }
        if (this.options.lazyHeaders) {
            this.modules.lazyHeaders = require('markdown-it-lazy-headers');
        }
        if (this.options.emoji) {
            this.modules.emoji = require('markdown-it-emoji');
        }
        if (this.options.expandTabs) {
            this.modules.expandTabs = require('markdown-it-expand-tabs');
        }
        if (this.options.abbr) {
            this.modules.abbr = require('markdown-it-abbr');
        }
        if (this.options.anchor) {
            this.modules.anchor = require('markdown-it-anchor');
        }
        if (this.options.attrs) {
            this.modules.attrs = require('markdown-it-attrs');
        }
        if (this.options.deflist) {
            this.modules.deflist = require('markdown-it-deflist');
        }
        if (this.options.footnote) {
            this.modules.footnote = require('markdown-it-footnote');
        }
        if (this.options.highlightjs) {
            this.modules.highlightjs = require('markdown-it-highlightjs');
        }
        if (this.options.ins) {
            this.modules.ins = require('markdown-it-ins');
        }
        if (this.options.mark) {
            this.modules.mark = require('markdown-it-mark');
        }
        if (this.options.sub) {
            this.modules.sub = require('markdown-it-sub');
        }
        if (this.options.sup) {
            this.modules.sup = require('markdown-it-sup');
        }
        if (this.options.mermaid) {
            this.modules.mermaid = require('markdown-it-mermaid-less');
        }
    }
    getRenderer() {
        const mdOptions = {
            html: true,
            langPrefix: 'highlight ',
            linkify: this.options.linkify,
        };
        const renderer = new this.modules.MarkdownIt({
            html: true,
            linkify: true,
            typographer: true,
        });
        if (this.options.taskLists) {
            renderer.use(this.modules.taskLists, { enabled: true, label: true });
        }
        if (this.options.lazyHeaders) {
            renderer.use(this.modules.lazyHeaders, {});
        }
        if (this.options.emoji) {
            renderer.use(this.modules.emoji, { shortcuts: {} });
        }
        if (this.options.expandTabs) {
            renderer.use(this.modules.expandTabs, { tabWidth: 4 });
        }
        if (this.options.abbr) {
            renderer.use(this.modules.abbr, {});
        }
        if (this.options.anchor) {
            renderer.use(this.modules.anchor, {});
        }
        if (this.options.attrs) {
            renderer.use(this.modules.attrs, {});
        }
        if (this.options.deflist) {
            renderer.use(this.modules.deflist, {});
        }
        if (this.options.footnote) {
            renderer.use(this.modules.footnote, {});
        }
        if (this.options.highlightjs) {
            renderer.use(this.modules.highlightjs, {});
        }
        if (this.options.ins) {
            renderer.use(this.modules.ins, {});
        }
        if (this.options.mark) {
            renderer.use(this.modules.mark, {});
        }
        if (this.options.sub) {
            renderer.use(this.modules.sub, {});
        }
        if (this.options.sup) {
            renderer.use(this.modules.sup, {});
        }
        if (this.options.mermaid) {
            renderer.use(this.modules.mermaid, this.mermaidOptions);
        }
        return renderer;
    }
}
exports.MarkdownRender = MarkdownRender;
